inst/tests/tests.Rraw

```diff
@@ -18330,3 +18330,51 @@ if (test_bit64) {
   apple = data.table(id = c("a", "b", "b"), time = c(1L, 1L, 2L), y = i64v[1:3])
   test(2248, dcast(apple, id ~ time, value.var = "y"), data.table(id = c('a', 'b'), `1` = i64v[1:2], `2` = i64v[4:3], key='id'))
 }
+
+# 5885 implement frev
+d = c(NA, NaN, Inf, -Inf)
+test(2249.00, frev(c(FALSE, NA),    copy=TRUE), c(NA, FALSE))
+test(2249.01, frev(c(0L, NA),       copy=TRUE), c(NA, 0L))
+test(2249.02, frev(d,               copy=TRUE), c(-Inf, Inf, NaN, NA))
+test(2249.03, frev(c(NA, 1, 0+2i),  copy=TRUE), c(0+2i, 1, NA))
+test(2249.04, frev(as.raw(0:1),     copy=TRUE), as.raw(1:0))
+test(2249.05, frev(NULL,            copy=TRUE), NULL)
+test(2249.06, frev(character(5),    copy=TRUE), character(5))
+test(2249.07, frev(integer(0),      copy=TRUE), integer(0))
+test(2249.08, frev(list(1, "a"),    copy=TRUE), list("a", 1))
+test(2249.09, frev(c(0L, NA),       copy=FALSE), c(NA, 0L))
+test(2249.10, frev(d,               copy=FALSE), c(-Inf, Inf, NaN, NA))
+test(2249.11, frev(c(NA, 1, 0+2i),  copy=FALSE), c(0+2i, 1, NA))
+test(2249.12, frev(as.raw(0:1),     copy=FALSE), as.raw(1:0))
+test(2249.13, frev(NULL,            copy=FALSE), NULL)
+test(2249.14, frev(character(5),    copy=FALSE), character(5))
+test(2249.15, frev(integer(0),      copy=FALSE), integer(0))
+test(2249.16, frev(list(1, "a"),    copy=FALSE), list("a", 1))
+test(2249.17, frev(1:1e2,           copy=TRUE),  rev(1:1e2))
+# copy arguments
+x = 1:3
+test(2249.21, {frev(x, copy=TRUE);  x}, 1:3)
+test(2249.22, {frev(x, copy=FALSE); x}, 3:1)
+# levels
+f = as.factor(letters)
+test(2249.31, frev(f, copy=TRUE), rev(f))
+test(2249.32, frev(as.IDate(1:10), copy=TRUE), as.IDate(10:1))
+test(2249.33, frev(as.IDate(1:10), copy=TRUE), as.IDate(10:1))
+# names
+x = c(a=1L, b=2L, c=3L)
+test(2249.41, frev(x, copy=TRUE), rev(x))
+test(2249.42, frev(x, copy=FALSE), x)
+# attributes
+x = structure(1:10, class = c("IDate", "Date"), att = 1L)
+test(2249.51, attr(frev(x, copy=TRUE), "att"), 1L)
+test(2249.52, attr(frev(x, copy=FALSE), "att"), 1L)
```

Comment:
Note that this will break e.g. the `gregexpr()` output, where the attributes are created to match the actual object order. Nothing we can do here besides document the potential issue, I think.

Reply:
True, but this problem also appears with `base::rev`?

Reply:
OK, let's just document that we copy base::rev behavior for attributes?

Reply:
Hmm but `rev()` drops attributes?

```
names(attributes(rev(structure(1:10, class = c("IDate", "Date"), att = 1L))))
# [1] "class"

names(attributes(rev(regexpr("[a-m]", letters))))
# NULL
```

Reply:
Good spot, I just checked with `names` and `class` but apparently those are special attributes.

Reply:
I added details in the docs about attributes being retained.

Reply:
I think we should rather match `base::rev()` behavior and strip (non-"special") attributes. As mentioned this can have subtly broken implications -- user wanting to keep attributes around should be responsible for how to do so.

Reply:
Not sure I'm on the same side here. Retaining attributes with all operations on a data.table is an often mentioned feature, users would like.

I know that it's not user-facing now, but what would be the behavior for `setrev` for stripping attributes? Strip them as a side effect or keep them?

Reply:
We drop attributes now on `frev`. For `setrev` we still keep them but can change if wanted.

For the special case of zero length vectors, attributes are still copied (`rev` behaves here the same).